### PR TITLE
[Geo] Refactor Point Field Mappers

### PR DESCRIPTION
**Pull request by nknize**
This PR refactors the following:

  * `GeoPointFieldMapper` and `PointFieldMapper` to `AbstractPointGeometryFieldMapper` derived from `AbstractGeometryFieldMapper`.
  * `.setupFieldType` moved up to `AbstractGeometryFieldMapper`
  * lucene indexing moved up to `AbstractGeometryFieldMapper.parse`
  * new `addStoredFields`, `addDocValuesFields` abstract methods for implementing stored field and doc values field indexing in the concrete field mappers

This refactor is the next phase for setting up a framework for extending spatial field mapper functionality in x-pack.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/AbstractGeometryFieldMapper.java [hunk 1]
```diff
@@ -54,7 +60,23 @@
         public static final Explicit<Boolean> IGNORE_Z_VALUE = new Explicit<>(true, false);
     }
 
-    public abstract static class Builder<T extends Builder, Y extends AbstractGeometryFieldMapper>
+    /**
+     * Interface representing an preprocessor in geo-shape indexing pipeline
+     */
```
Comment (iverase) [inline server/src/main/java/org/elasticsearch/index/mapper/AbstractGeometryFieldMapper.java:65]: nit: 
s/geo-shape/geometry

File: server/src/main/java/org/elasticsearch/index/mapper/AbstractGeometryFieldMapper.java [hunk 2]
```diff
@@ -54,7 +60,23 @@
...
+    public interface Indexer<Parsed, Processed> {
+        Processed prepareForIndexing(Parsed geometry);
+        Class<Processed> processedClass();
+        List<IndexableField> indexShape(ParseContext context, Processed shape);
+    }
+
+    /**
+     * interface representing parser in geo shape indexing pipeline
+     */
```
Comment (iverase) [inline server/src/main/java/org/elasticsearch/index/mapper/AbstractGeometryFieldMapper.java:74]: nit:
s/geo-shape/geometry

File: server/src/main/java/org/elasticsearch/index/mapper/AbstractPointGeometryFieldMapper.java [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+
+import java.io.IOException;
+import java.text.ParseException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+/** Base class for GeoPointFieldMappers */
+import static org.elasticsearch.index.mapper.TypeParsers.parseField;
```
Comment (iverase) [inline server/src/main/java/org/elasticsearch/index/mapper/AbstractPointGeometryFieldMapper.java:39]: nit: javadoc is too generic? maybe:

/** Base class for for spatial fields that only support indexing points */

Comment (nknize) [reply server/src/main/java/org/elasticsearch/index/mapper/AbstractPointGeometryFieldMapper.java:39]: :+1:

File: server/src/main/java/org/elasticsearch/index/mapper/AbstractGeometryFieldMapper.java (hunk 1 above)
Comment (nknize) [reply server/src/main/java/org/elasticsearch/index/mapper/AbstractGeometryFieldMapper.java:65]: :+1:

File: server/src/main/java/org/elasticsearch/index/mapper/AbstractGeometryFieldMapper.java (hunk 2 above)
Comment (nknize) [reply server/src/main/java/org/elasticsearch/index/mapper/AbstractGeometryFieldMapper.java:74]: :+1:

**Pull request merged**